### PR TITLE
Proxy: Implement TLS conditional accept more like TLS conditional connect.

### PR DESCRIPTION
**Pull request by briansmith**
Clean up the accept side of the TLS configuration logic.

Signed-off-by: Brian Smith <brian@briansmith.org>

### CODE REVIEW

File: proxy/src/connection.rs [hunk 1]
```diff
@@ -133,32 +133,31 @@ impl BoundPort {
...
-                                });
-                            return Either::A(f);
-                        } else {
-                            // No valid TLS configuration.
-                            tls::ReasonForNoTls::NoConfig
-                        }
-                    } else {
-                        tls::ReasonForNoTls::Disabled
+
+                    let conn = match tls::current_connection_config(&tls) {
```
Comment (hawkw) [inline proxy/src/connection.rs]: This is much cleaner --- I take back anything negative I said about the `Conditional` type. :)

File: proxy/src/connection.rs [hunk 2]
```diff
@@ -111,7 +111,7 @@ impl BoundPort {
     // TLS when needed.
     pub fn listen_and_fold<T, F, Fut>(
         self,
-        tls: Option<(tls::Identity, tls::ServerConfigWatch)>,
+        tls: tls::ConditionalConnectionConfig<tls::ServerConfigWatch>,
```
Comment (hawkw) [inline proxy/src/connection.rs]: NIT/TIOLI: this type name is pretty long; maybe the `tls` module should expose aliases like
```rust
pub type ConditionalServerConfig = ConditionalConnectionConfig<ServerConfigWatch>;
pub type ConditionalClientConfig = ConditionalConnectionConfig<ClientConfigWatch>;
```

File: proxy/src/lib.rs [hunk 3]
```diff
@@ -259,9 +259,16 @@ where
                 config.inbound_router_capacity,
                 config.inbound_router_max_idle_age,
             );
+            let tls_settings = match &config.tls_settings {
+                Some(settings) => Conditional::Some(tls::ConnectionConfig {
+                    identity: settings.service_identity.clone(),
+                    config: tls_server_config
+                }),
+                None => Conditional::None(tls::ReasonForNoTls::NoConfig),
```
Comment (hawkw) [inline proxy/src/lib.rs]: TIOLI: Seems like this is a common enough pattern that a function like 
```rust
impl<C, R> Conditional<C, R> {
    // ... 
    pub fn from_option(opt: Option<C>, reason: R) -> Conditional<C, R> {
        match opt {
            Some(c) => Conditional::Some(c),
            None => Conditional::None(reason),
        }
    }
}
```
would be nice?

(Or even an extension trait that gives options a `conditional_or()` conversion?)

File: proxy/src/connection.rs [hunk 4]
```diff
@@ -133,32 +133,31 @@ impl BoundPort {
...
+                            // TODO: use `config.identity` to differentiate
+                            // between TLS that the proxy should terminate vs.
+                            // TLS that should be passed through.
+                            let f = tls::Connection::accept(socket, config.config)
+                                .map(move |tls| Connection::tls(tls));
+                            Either::A(f)
+                        },
+                        Conditional::None(why_no_tls) => {
+                            let f = future::ok(socket)
+                                .map(move |plain| Connection::plain(plain, why_no_tls));
```
Comment (hawkw) [inline proxy/src/connection.rs]: NIT: I think this can be just
```rust
let f = future::ok(Connection::plain(socket, why_no_tls));
```

**Comment by hawkw** [comment]
BTW, it looks like the reason your CI build is broken is that some of the telemetry tests don't expect the `tls="no_config"` label that's present after your changes. This should be simple to fix.

File: proxy/src/connection.rs (hunk 4 above)
Comment (hawkw) [reply proxy/src/connection.rs]: Ah, that makes sense. To me, reading this just makes me think "Ah, this case used to be more complex and some intermediate combinators were removed" or something, but I understand the rationale.

File: proxy/src/lib.rs (hunk 3 above)
Comment (briansmith) [reply proxy/src/lib.rs]: FYI, I just looked at implementing `map` and having it doesn't reduce the amount of code (in terms of number of lines), so I threw it away.

**Comment by briansmith** [comment]
> BTW, it looks like the reason your CI build is broken is that some of the telemetry tests don't expect the tls="no_config" label that's present after your changes. This should be simple to fix.

The problem was that I was reporting the lack of a `tls::Settings` in config.rs as `NoConfig` instead of `Disabled`. I changed `src/config.rs` to use `Conditional` to reduce the chances for such an error and that fixed it.

**Pull request merged**